### ISSUE TITLE
Implement libtorch backend with autodiff support

### ISSUE THREAD

**Issue by shinaoka** [shinaoka/tensor4all-rs#139]
## Goal
Add a **libtorch (tch-rs) storage backend** to `tensor4all-tensorbackend` so that tensor operations in the core stack can be executed on PyTorch tensors and participate in **autograd** (backward + higher-order).

This repo is in **early development** (see `AGENTS.md`), so breaking changes are acceptable if they simplify the architecture.

---

## Current Status (as of 2026-01-15)
- `tensor4all-tensorbackend` already provides:
  - `Storage` as a concrete enum (Dense/Diag × f64/Complex64)
  - `AnyScalar` as a value enum (f64/Complex64)
  - linear algebra backend dispatch for SVD/QR (`backend-faer`, `backend-lapack`)
- `tensor4all-core` is built on top of this concrete `Storage` API (permute/contract/sum/scale/etc.).
- There is **no** `backend-libtorch` feature and no `tch` dependency yet.

---

## Scope (v1)
- **CPU only**.
- **Feature parity is important**: the torch backend must support the same value-level feature set as current mdarray backend:
  - Dense `f64`
  - Dense `Complex64`
  - Diag `f64`
  - Diag `Complex64`

### Autodiff scope (v1)
- Autodiff support is required at least for **real (f64) dense tensors**.
- **Complex autograd approach (chosen): Option A (native complex)**
  - Use PyTorch complex tensors (via `tch`) for Complex64 storage.
  - If an operation needed by tensor4all is not supported for complex autograd upstream, the implementation must:
    - fail fast with a clear error message describing the unsupported op/dtype, OR
    - implement an explicit alternative with tests.
  - Do **not** silently fall back to real/imag split in v1 (track as separate issue if needed).

### Diag representation scope (v1)
Diag is a first-class feature; pick and document one approach:
- **Option A (preserve Diag semantics)**: represent diag storage explicitly and implement specialized ops (better perf, more work).
- **Option B (densify with explicit trade-off)**: support Diag by converting to Dense in torch backend for now (correctness-first; perf follow-up).

---

## Non-goals (v1)
- GPU support
- Runtime backend dispatch (keep compile-time selection)
- Language bindings packaging (Julia/Python split packages) — track separately once Rust core works

---

### Compile-time backend selection
Introduce a **storage backend selection feature** (name bikesheddable):

- `backend-mdarray` (default): current implementation
- `backend-libtorch`: new implementation (depends on `tch`)

Expose a single `Storage` type at the crate root via cfg:

```rust
// tensor4all-tensorbackend/src/lib.rs
#[cfg(feature = "backend-libtorch")]
pub type Storage = torch::TorchStorage;

#[cfg(not(feature = "backend-libtorch"))]
pub type Storage = mdarray::MdarrayStorage;
```

### Common surface area
Define a minimal common API that `tensor4all-core` needs (either as a trait + type alias, or as identical inherent methods on both concrete types). Must cover at least:
- creation from raw data / conversion to `DTensor<f64>` and `DTensor<Complex64>`
- `permute`
- `conj`
- contraction along axes (tensordot-style)
- `sum`, `scale`, `try_add`, `try_sub`, `axpby`

### Autodiff API
For `backend-libtorch`, expose autograd-oriented hooks:
- `requires_grad`, `set_requires_grad`
- `grad()`
- `backward()`
- `backward_with_graph()` or `grad_with_graph()` (for higher-order)

For `backend-mdarray`, these are **no-ops** returning `false`/`None`.

---

### Phase 1 — Refactor for backend selection (no behavior change)
- [ ] Move the current `Storage`/`AnyScalar` implementation into `src/mdarray/` module
- [ ] Add a backend-neutral `Storage` facade (type alias + re-exports)
- [ ] Update `tensor4all-core` to use the facade (should mostly be re-export compatible)
- [ ] Ensure **all existing tests pass** with default features

### Phase 2 — Add torch backend (CPU) with parity (Dense/Diag × f64/Complex64)
- [ ] Add optional dependency `tch` behind `backend-libtorch`
- [ ] Implement `TorchStorage`:
  - [ ] Dense f64
  - [ ] Dense Complex64 (native complex tensors)
  - [ ] Diag f64 (via Option A or B)
  - [ ] Diag Complex64 (via Option A or B)
- [ ] Implement required operations (permute/contract/conj/sum/scale/add/sub/axpby/convert)
- [ ] Add unit tests in `tensor4all-tensorbackend` validating parity vs mdarray backend for:
  - [ ] Dense f64
  - [ ] Dense Complex64
  - [ ] Diag f64
  - [ ] Diag Complex64

### Phase 3 — Autodiff
- [ ] Implement `requires_grad`/`grad`/`backward` for TorchStorage (at least Dense f64)
- [ ] Add tests:
  - [ ] scalar gradient (x^2)
  - [ ] tensor contraction gradient (e.g. matmul-like contraction)
  - [ ] double backward (at least 1 small example)
- [ ] Add complex gradient tests for operations that are supported under the native complex approach.

### Phase 4 — Integration & packaging (follow-up)
- [ ] Expose backend selection to C API
- [ ] Plan Python/Julia binary distribution strategy (likely separate packages)

---

## Acceptance Criteria
- Default build/test remains green:
  - `cargo test -p tensor4all-tensorbackend`
  - `cargo test -p tensor4all-core`
- Torch backend builds and its tests pass:
  - `cargo test -p tensor4all-tensorbackend --features backend-libtorch`
  - (optional stretch) `cargo test -p tensor4all-core --features backend-libtorch`
- Parity tests cover Dense/Diag × f64/Complex64
- Autodiff tests validate correct gradients for:
  - [ ] Dense f64
  - [ ] Dense Complex64 (native complex; limited to ops that are upstream-supported)

---

## Notes / Risks
- `tch`/libtorch toolchain friction (platform-specific) — keep feature-gated and CPU-only initially.
- Complex autograd support in PyTorch exists but may be incomplete depending on ops; under Option A we must be explicit about what is/ isn't supported and fail loudly when unsupported.
- Diag may be densified initially for correctness; follow-up can restore specialized diag performance.

**Comment by shinaoka** [comment]
## Progress Update (2026-01-17)

### Recent Changes Affecting This Issue

PR #141 introduces significant architectural changes that affect the torch backend integration:

1. **New `mdarray-einsum` crate** - Hyperedge-aware tensor contraction optimizer
   - Provides optimized einsum implementation with greedy contraction path optimization
   - Supports hyperedges (indices appearing in 3+ tensors)
   - `TypedTensor` enum wraps either f64 or Complex64 tensors with automatic type promotion
   - Backend-agnostic design via `MatMul<T>` trait

2. **Refactored contraction in `tensor4all-core`**
   - `contract.rs` and `contraction_opt.rs` merged (DRY principle)
   - Uses `mdarray-einsum` for optimized contraction paths
   - Preserves storage type: F64 inputs → F64 output, any C64 input → C64 output

### Implications for Torch Backend

The `MatMul<T>` trait in `mdarray-einsum` provides a natural integration point:

```rust
pub trait MatMul<T: Scalar> {
    fn matmul(a: &Slice<T, D2, impl Layout>, b: &Slice<T, D2, impl Layout>) -> Tensor<T, D2>;
}
```

**Suggested approach:**
- Implement `MatMul<f64>` and `MatMul<Complex64>` for a `TorchBackend` struct
- This allows reusing the existing contraction path optimization logic
- The optimizer is backend-agnostic; only the actual matrix multiplication needs torch implementation

### Updated Implementation Considerations

**Phase 2 adjustments:**
- Consider implementing torch backend at the `MatMul` trait level first
- This provides immediate benefit for contraction operations
- Storage-level integration can follow

**New integration point:**
```rust
// Example: torch matmul backend
pub struct TorchMatMul;

impl MatMul<f64> for TorchMatMul {
    fn matmul(a: &Slice<f64, D2, impl Layout>, b: &Slice<f64, D2, impl Layout>) -> Tensor<f64, D2> {
        // Convert to tch::Tensor, matmul, convert back
    }
}
```

This layered approach allows:
1. Quick wins with contraction optimization using torch matmul
2. Gradual storage-level integration as needed
3. Autodiff through the matmul operation

**Comment by shinaoka** [comment]
## Architecture Note: Einsum Trait Instead of MatMul

After reviewing the PyTorch codebase (`extern/pytorch`), I realized the current approach of using `MatMul<T>` trait as the backend abstraction point is not ideal.

### Current Approach (Not Optimal)
- `mdarray-einsum` depends on `MatMul<T>` trait from `mdarray-linalg`
- Einsum is implemented by decomposing into matmul operations
- This loses the opportunity to use native backend optimizations

### Better Approach: Einsum Trait in tensorbackend

PyTorch has native `torch.einsum()` with:
- Batch dimension support
- Contraction path optimization (via opt-einsum)
- Automatic differentiation support

**Proposed Design:**

```rust
// crates/tensor4all-tensorbackend/src/einsum.rs

/// Backend trait for einsum operations.
pub trait EinsumBackend {
    type Tensor;
    type Scalar;
    
    /// Perform einsum with string notation: "ij,jk->ik"
    fn einsum_str(
        &self,
        equation: &str,
        operands: &[&Self::Tensor],
    ) -> Self::Tensor;
    
    /// Perform einsum with explicit index labels
    fn einsum(
        &self,
        inputs: &[(&[usize], &Self::Tensor)],
        output_ids: &[usize],
    ) -> Self::Tensor;
}
```

**Backend Implementations:**

1. **mdarray backend**: Uses current `mdarray-einsum` logic (sumproduct_pair, hyperedge optimizer)
2. **torch backend**: Calls `tch::Tensor::einsum()` directly, leveraging PyTorch's native implementation

### Benefits

1. **Autodiff**: PyTorch's einsum natively supports autograd - we get this for free
2. **Batch support**: PyTorch handles batched einsum via ellipsis notation
3. **Optimization**: PyTorch can use opt-einsum for contraction path optimization
4. **Simpler integration**: No need to decompose operations - just call the native function

### Implementation Plan

1. Define `EinsumBackend` trait in `tensor4all-tensorbackend`
2. Implement for mdarray (wrapping current `mdarray-einsum`)
3. Implement for torch (direct `Tensor::einsum()` call)
4. Update `tensor4all-core` to use `EinsumBackend` instead of direct einsum calls

This aligns with the goal of keeping the torch backend simple while getting full autodiff support.
